### PR TITLE
Filter workspaces based on user permissions

### PR DESCRIPTION
**Pull request by waxlamp**
This PR implements one permissions-based filter for a part of the API. The `GET /workspaces` endpoint now filters the returned workspaces based on the logged in user and properties of the workspaces themselves.

This PR is also a prototype of sorts: what I've done here works but might require some refactoring after we start to fill in more permission-based changes for the other endpoints.

The workspaces endpoint is special in that it's not all or nothing: the full list of workspaces has to come from the database, and then we evaluate what to send as the response. I suspect other endpoints will have a simpler gating mechanism which results in a 200 or a 40x error depending on the user.

This PR adds a new semantics of workspace creation depending on who (if anyone) is logged in: a logged-in user creates a non-public workspace owned by them; a logged-out user creates a public workspace owned by nobody. Only an admin would be able to delete such workspaces, so this is likely the wrong thing to do. What would be better? I suspect it would simply be to disallow logged-out users from creating workspaces.

TODO:
- [x] write a script to upgrade the workspace mapping collection to have the right schema (with proper filled in data)
- [x] update the workspace creation endpoint to fill in correct metadata for the new workspace
- [x] ~use `Optional[User]` as the type for the `owner` field of `Workspace`~
- [x] prevent non-logged-in user from creating workspaces
- [x] ~add tests~
- [x] perform DB migration (before merge)

### CODE REVIEW

File: multinet/api.py
Comment (JackWilb) [inline multinet/api.py:160]: This means that users that are not logged in can make workspaces. Is this what we resolved to do? Also, @waxlamp, you mentioned that you could make the owner an optional string, instead of just blank, which sounds like a good idea

Comment (waxlamp) [reply multinet/api.py:160]: > This means that users that are not logged in can make workspaces. Is this what we resolved to do?

This is exactly what we need to discuss; see the PR description.

> Also, @waxlamp, you mentioned that you could make the owner an optional string, instead of just blank, which sounds like a good idea

That does make more sense from a typing standpoint, but from an execution standpoint, there's a nice advantage to using a blank string for "no owner": no real user has a blank string as their `sub` value, so it leaves all the code for processing users nice and simple. OTOH, the type safety is pretty attractive. I will probably make this change but if there are opposing viewpoints, now would be a good time to hear them.

Comment (JackWilb) [reply multinet/api.py:160]: > > This means that users that are not logged in can make workspaces. Is this what we resolved to do?
> 
> This is exactly what we need to discuss; see the PR description.

Understood. I think we shouldn't allow users that are not logged in to create workspaces. This could lead to abuse, eventually, and it makes sense that every workspace should have an owner. A workspace not having an owner could lead to confusion, where users make workspaces when they're not logged in, then they become public and they're unable to edit them.

I welcome any counter arguments though

> > Also, @waxlamp, you mentioned that you could make the owner an optional string, instead of just blank, which sounds like a good idea
> 
> That does make more sense from a typing standpoint, but from an execution standpoint, there's a nice advantage to using a blank string for "no owner": no real user has a blank string as their `sub` value, so it leaves all the code for processing users nice and simple. OTOH, the type safety is pretty attractive. I will probably make this change but if there are opposing viewpoints, now would be a good time to hear them.

I think we should make the change for better typing. I think making it an optional string will mean that we're less like to forget about the edge case where a workspace has no owner.

Comment (waxlamp) [reply multinet/api.py:160]: > > > This means that users that are not logged in can make workspaces. Is this what we resolved to do?
> > 
> > 
> > This is exactly what we need to discuss; see the PR description.
> 
> Understood. I think we shouldn't allow users that are not logged in to create workspaces. This could lead to abuse, eventually, and it makes sense that every workspace should have an owner. A workspace not having an owner could lead to confusion, where users make workspaces when they're not logged in, then they become public and they're unable to edit them.
> 
> I welcome any counter arguments though
> 

I agree that we should just not allow it for unlogged-in users. @AlmightyYakob, any objection?

I will add a commit to handle this case; it'll be a good prototype use case for a `401 Unauthorized` response.

> > > Also, @waxlamp, you mentioned that you could make the owner an optional string, instead of just blank, which sounds like a good idea
> > 
> > 
> > That does make more sense from a typing standpoint, but from an execution standpoint, there's a nice advantage to using a blank string for "no owner": no real user has a blank string as their `sub` value, so it leaves all the code for processing users nice and simple. OTOH, the type safety is pretty attractive. I will probably make this change but if there are opposing viewpoints, now would be a good time to hear them.
> 
> I think we should make the change for better typing. I think making it an optional string will mean that we're less like to forget about the edge case where a workspace has no owner.

If we're not allowing ownerless workspaces, then the `owner` field *should* be `str` rather than `Optional[str]` no?

Comment (AlmightyYakob) [reply multinet/api.py:160]: > I agree that we should just not allow it for unlogged-in users. @AlmightyYakob, any objection?

@waxlamp I agree, only logged in user should be able to create a workspace. 

> If we're not allowing ownerless workspaces, then the `owner` field _should_ be `str` rather than `Optional[str]` no?

I think so, I don't see any particular benefit to making it optional. However, this means that we'll have to do some migration on existing workspaces. Maybe we could set them to some default user that we create (and have access to), and then we can assign them to whoever they belong to.

Comment (JackWilb) [reply multinet/api.py:160]: I agree with Jake on that last point

Comment (waxlamp) [reply multinet/api.py:160]: > Maybe we could set them to some default user that we create (and have access to), and then we can assign them to whoever they belong to.

Since we don't have an in-app admin console yet, I don't see the value in assigning these to an existing user, as opposed to just leaving the string blank, and setting the workspaces to be public. I wrote a migration script (2d1491c143d10d10b0b6a07bb26209f6590fc36d in this PR) to do just that. Once the actual owners of these workspaces have accounts, we can re-assign them manually.

**Comment by AlmightyYakob** [comment]
It would also be good to have tests for this new behavior, as it adds a decent amount of complexity and it'll be good to keep these things straight.

File: multinet/api.py
Comment (waxlamp) [reply multinet/api.py]: Added a 401 error for logged-out users trying to create a workspace: see dbabdbc.

**Comment by waxlamp** [comment]
This is now ready for final review.

**Comment by JackWilb** [comment]
I'm having an issue. When I'm not logged in, the `/api/workspaces` endpoint is returning me invalid JSON, instead of an empty list. It returns:

```
[
```

I'm also struggling to sign in using the api endpoints on the preview app. It's giving me an error about the redirect URI

**Comment by AlmightyYakob** [comment]
> I'm having an issue. When I'm not logged in, the `/api/workspaces` endpoint is returning me invalid JSON, instead of an empty list. It returns:
> 
> ```
> [
> ```


This sounds like another instance of #347. Did you encounter this error when testing the app locally? Or were you using the preview app?

> 
> I'm also struggling to sign in using the api endpoints on the preview app. It's giving me an error about the redirect URI

This is probably due to us having to explicitly register the redirect URLs on Google's side for OAuth. I'm guessing @waxlamp didn't add the preview app to the allowed list in the dev console. We don't have a great process for this yet.

**Comment by JackWilb** [comment]
> This sounds like another instance of #347. Did you encounter this error when testing the app locally? Or were you using the preview app?

This was on the preview app. It definitely seems related to #347.

> This is probably due to us having to explicitly register the redirect URLs on Google's side for OAuth. I'm guessing @waxlamp didn't add the preview app to the allowed list in the dev console. We don't have a great process for this yet.

I figured. Is there a way to test the login without adding the redirct URI to google? Even locally, right now?

File: multinet/api.py
Comment (AlmightyYakob) [inline multinet/api.py:49]: I think this would be better placed outside of `get_workspaces` (and probably renamed to something like `workspace_readable`), as nothing in this function relies on anything besides the input parameters. 

Alternatively, we could make a general `workspace_permissions` function (name not withstanding), which would take a workspace and a user, but would instead return all of the permissions the user has on that workspace, instead of just read, just write, etc. This would have a new signature that describes a user's permissions on a workspace, something like:

```
{
    "owner": bool,
    "reader": bool,
    "writer": bool,
    "maintainer": bool
}
```

Then below, instead of `can_read(user, w)`, you'd replace it with `workspace_permissions(user, w)["reader"]`.

I'd also think that if we were to create this function, we'd put it in either `db.py` or `util.py`.

File: multinet/api.py
Comment (AlmightyYakob) [inline multinet/api.py:31]: Is this left over from debugging?

File: multinet/db.py
Comment (AlmightyYakob) [inline multinet/db.py:129]: This may not be strictly related to this PR, but I think here we should wait to insert the workspace document into the mapping collection until after the database creation has succeeded. The way that things are currently implemented, we could have dangling entries in the workspace mapping collection. I'm okay with addressing this in a separate issue if that's what we want to do.

File: multinet/types.py
Comment (AlmightyYakob) [inline multinet/types.py:11]: To clarify, this maps to the `sub` field of a user? If so, it might be good to add a comment here to explain that.

**Review by jjnesbitt — CHANGES_REQUESTED** [review]

**Comment by AlmightyYakob** [comment]
> > This is probably due to us having to explicitly register the redirect URLs on Google's side for OAuth. I'm guessing @waxlamp didn't add the preview app to the allowed list in the dev console. We don't have a great process for this yet.
> 
> I figured. Is there a way to test the login without adding the redirct URI to google? Even locally, right now?

I don't think so, as the login always calls out to Google. However, we can temporarily add the preview app to the allowed origins in the Google dev console, for testing this. Any objections @waxlamp?

**Comment by AlmightyYakob** [comment]
> > This sounds like another instance of #347. Did you encounter this error when testing the app locally? Or were you using the preview app?
> 
> I don't think so--this happens because of how we turn an iterable into a JSON response, [here](https://github.com/multinet-app/multinet-server/blob/master/multinet/util.py#L65-L74). If an exception is raised during the "meat" of this function, you'll get a malformed JSON response. It is not a great way to do this, so I welcome any ideas for improving it.

My bad, what I meant was that this same issue with the `generate` function that is documented in #347. We should probably address the bug with the `generate` function separately from addressing #347, as it's a more general bug.

File: devops/scripts/ensure_workspace_metadata.py
Comment (JackWilb) [inline devops/scripts/ensure_workspace_metadata.py:5]: mypy_extensions are not in the Pipfile which causes errors on fresh installs. It may be worth adding it in, even as a dev dependency

**Review by JackWilb — COMMENTED** [review]
Just a little note here in this review that I noticed on a fresh install.

Also, I think it would definitely be worth testing the login and ownership before we push, I'm not sure how best to do that. We could set the preview to point at the test arango db instance and add in the preview for the google logins. That means the main database wouldn't be affected until we're ready to go.

We'll also need to think about using sensitive/not public datasets on our applications and how to handle those logins. Would users just need to log into multinet first and be (at least) a reader of the workspace?

File: multinet/user.py
Comment (waxlamp) [reply multinet/user.py]: See 8754b2a73aaab43a5135b9a4eb291e83dc117ee9.

File: multinet/user.py
Comment (waxlamp) [reply multinet/user.py]: In discussions we settled on `user_from_cookie()` for the name of this function (eliminating the imperative-sounding verb from the function name).

See c58546db94ec3da1883669e01e37ad4652715b00.

File: multinet/api.py
Comment (waxlamp) [reply multinet/api.py]: All excellent suggestions. See 8cd1fb5052075ff635d704726ae8a44997c7d60b (please make sure I addressed all points).

File: multinet/api.py
Comment (waxlamp) [reply multinet/api.py]: See 8cd1fb5052075ff635d704726ae8a44997c7d60b.

File: multinet/db.py
Comment (waxlamp) [reply multinet/db.py]: See 255bfb6d356cc60a89f78d13e8d808ab82d49189.

In general, we may want to write a sort of fsck script to validate our server data state. See #398.

File: multinet/types.py
Comment (waxlamp) [reply multinet/types.py]: See ca96a922d26c192cce8fbcada77024cca62196e6.

File: devops/scripts/ensure_workspace_metadata.py
Comment (waxlamp) [reply devops/scripts/ensure_workspace_metadata.py:5]: Good catch. For that matter, `mypy` isn't listed in `Pipfile` either. I will send this update on a separate branch, so as not to conflict with the content changes in this branch.

**Comment by waxlamp** [comment]
> Also, I think it would definitely be worth testing the login and ownership before we push, I'm not sure how best to do that. We could set the preview to point at the test arango db instance and add in the preview for the google logins. That means the main database wouldn't be affected until we're ready to go.

I agree with you and @AlmightyYakob both that testing is really needed for this type of addition. However, since our testing infrastructure requires a bit of love, I'm going to defer that to after this PR is merged. In particular, there's a little bit of cleanup we can do, and some skipped tests that need to be unskipped, and we need to institute our policy of testing with the testing DB. That all seems like too much for this branch, so we'll live on the edge a bit until that's done.

> We'll also need to think about using sensitive/not public datasets on our applications and how to handle those logins. Would users just need to log into multinet first and be (at least) a reader of the workspace?

I'm not sure what you mean exactly--if there's sensitive data in the system, it would be owned by whoever uploaded it, and that user can grant permissions to view the workspace. We do need to make sure that people don't assume there's more security here than there actually is; I've been getting some advice from our legal folks on how this can work (most likely by posting a notice that explains that you need to remain responsible for your data, etc.).

File: devops/scripts/ensure_workspace_metadata.py
Comment (waxlamp) [reply devops/scripts/ensure_workspace_metadata.py:5]: See #399.

File: multinet/api.py
Comment (waxlamp) [reply multinet/api.py:49]: In our permissions model, each role includes implicitly the roles that are "weaker"; that is, an owner of a workspace is also, by definition, a maintainer, writer, and reader. Since that's the case, and our needs are for functions that can filter by these roles, I'm going to create a collection of `-> bool` functions for each role, and create associated decorators to guard the use of API endpoints (in this case, we wouldn't use the decorator, since we're filtering all workspaces by that hypothetical `is_reader()` function).

I will push a commit shortly to do a bit of refactoring here.

**Comment by JackWilb** [comment]
> > Also, I think it would definitely be worth testing the login and ownership before we push, I'm not sure how best to do that. We could set the preview to point at the test arango db instance and add in the preview for the google logins. That means the main database wouldn't be affected until we're ready to go.
> 
> I agree with you and @AlmightyYakob both that testing is really needed for this type of addition. However, since our testing infrastructure requires a bit of love, I'm going to defer that to after this PR is merged. In particular, there's a little bit of cleanup we can do, and some skipped tests that need to be unskipped, and we need to institute our policy of testing with the testing DB. That all seems like too much for this branch, so we'll live on the edge a bit until that's done.

I meant even some manual testing that things are working since it's difficult to verify that this PR is working as intended without updating the database to the new permissions model and testing it with logins. I agree that the automated tests can be a follow up PR.

> > We'll also need to think about using sensitive/not public datasets on our applications and how to handle those logins. Would users just need to log into multinet first and be (at least) a reader of the workspace?
> 
> I'm not sure what you mean exactly--if there's sensitive data in the system, it would be owned by whoever uploaded it, and that user can grant permissions to view the workspace. We do need to make sure that people don't assume there's more security here than there actually is; I've been getting some advice from our legal folks on how this can work (most likely by posting a notice that explains that you need to remain responsible for your data, etc.).

I mean for an app like Jen's evo-bio. If she is sharing data that is not public on there, how does she grant access to people to use data from multinet that she uploaded? Is the flow: she uploads, adds the person she wants to show it to as a viewer, and then they have to be logged into multinet in the background while viewing her app?

File: multinet/api.py
Comment (waxlamp) [reply multinet/api.py:49]: See 282cbf4101ae3b4e0a221453ce3f82a6bc16f997.

File: multinet/api.py
Comment (waxlamp) [reply multinet/api.py]: Yup :)

See a9e328c4e2404460df452d0107acd500db50acac.

File: multinet/api.py
Comment (waxlamp) [reply multinet/api.py]: (I think I did address everything, so I'll resolve this conversation.)

**Comment by waxlamp** [comment]
> > > Also, I think it would definitely be worth testing the login and ownership before we push, I'm not sure how best to do that. We could set the preview to point at the test arango db instance and add in the preview for the google logins. That means the main database wouldn't be affected until we're ready to go.
> > 
> > 
> > I agree with you and @AlmightyYakob both that testing is really needed for this type of addition. However, since our testing infrastructure requires a bit of love, I'm going to defer that to after this PR is merged. In particular, there's a little bit of cleanup we can do, and some skipped tests that need to be unskipped, and we need to institute our policy of testing with the testing DB. That all seems like too much for this branch, so we'll live on the edge a bit until that's done.
> 
> I meant even some manual testing that things are working since it's difficult to verify that this PR is working as intended without updating the database to the new permissions model and testing it with logins. I agree that the automated tests can be a follow up PR. It's possible to test that workspaces are either public or hidden, but not that logging in reveals them to the owner of the workspace.
> 

Ah, gotcha.

I have been doing this type of manual testing locally. If you log in, create a few workspaces, and make some of them public, then log out (and refresh) you should see that the non-public workspaces are hidden from view.

(There is a bug where logging out doesn't refresh the workspace list, and that's something I should address either on this branch or in an immediate followup.)

Let me know if you're not able to build locally in order to test. I am able to do it, so there must be a way to make it work on your end too--let me know if you want to work it out together.

> > > We'll also need to think about using sensitive/not public datasets on our applications and how to handle those logins. Would users just need to log into multinet first and be (at least) a reader of the workspace?
> > 
> > 
> > I'm not sure what you mean exactly--if there's sensitive data in the system, it would be owned by whoever uploaded it, and that user can grant permissions to view the workspace. We do need to make sure that people don't assume there's more security here than there actually is; I've been getting some advice from our legal folks on how this can work (most likely by posting a notice that explains that you need to remain responsible for your data, etc.).
> 
> I mean for an app like Jen's evo-bio. If she is sharing data that is not public on there, how does she grant access to people to use data from multinet that she uploaded? Is the flow: she uploads, adds the person she wants to show it to as a viewer, and then they have to be logged into multinet in the background while viewing her app?

That's exactly the workflow as far as I know. In Jen's particular case, the data was always public, so anyone can view it in Trevo without being logged in. But if it were non-public data, then it would go exactly as you said.

We'll need to test this out in practice to see if it works properly.

**Comment by waxlamp** [comment]
> (There is a bug where logging out doesn't refresh the workspace list, and that's something I should address either on this branch or in an immediate followup.)

I just realized this is a *client* bug so of course I won't fix it on this branch.  See https://github.com/multinet-app/multinet-client/issues/103.

File: multinet/api.py
Comment (JackWilb) [inline multinet/api.py:48]: You refactored this, but left the old function here. Looks to me like this can be removed

File: multinet/auth/util.py
Comment (JackWilb) [inline multinet/auth/util.py:24]: Thinking ahead to when we implement different actions that will require logins (making tables and networks, for example), we'll probably want this to be a bit more general. Is it worth doing the work to make it general now or deferring it until we modify the other endpoints to require a log in?

**Review by JackWilb — CHANGES_REQUESTED** [review]
Just a couple notes, but this looks really close

File: multinet/auth/util.py
Comment (waxlamp) [reply multinet/auth/util.py:24]: Do you mean the error message or the whole function?

The error message should be fixed now; I'll take care of that shortly.

File: multinet/api.py
Comment (waxlamp) [reply multinet/api.py]: Good catch. See 4f2d28d18263d845646c3cd976d467943cf7d88e.

File: multinet/auth/util.py
Comment (waxlamp) [reply multinet/auth/util.py]: See 1387677bfceabc80985d05d29cea2d084ed45d74.

Comment (JackWilb) [reply multinet/auth/util.py]: Works for me

**Review by JackWilb — DISMISSED** [review]

**Review by JackWilb — DISMISSED** [review]

File: multinet/auth/util.py
Comment (AlmightyYakob) [inline multinet/auth/util.py:44]: Small thing, but I think as a matter of principle, we should check permissions from most common to least common. Since the role of `owner` is the least common, I'd suggest the following:
```suggestion
    return (
        perms["public"]
        or sub in perms["readers"]
        or sub in perms["writers"]
        or sub in perms["maintainers"]
        or perms["owner"] == sub
    )
```

**Review by jjnesbitt — CHANGES_REQUESTED** [review]

File: multinet/auth/util.py
Comment (AlmightyYakob) [inline multinet/auth/util.py:24]: Fixing my other suggestion, as it was out of date. Sorry it's so large, I wanted to include the import:

```suggestion
from typing import Any, Optional, Callable

from multinet.errors import Unauthorized
from multinet.types import Workspace
from multinet.auth.types import UserInfo
from multinet.user import current_user


# NOTE: unfortunately, it is difficult to write a type signature for this
# decorator. I've opened an issue to ask about this here:
# https://github.com/python/mypy/issues/9032.
def require_login(f: Callable) -> Callable:
    """Decorate an API endpoint to check for a logged in user."""

    def wrapper(*args: Any, **kwargs: Any) -> Any:
        user = current_user()
        if user is None:
            raise Unauthorized("You must be logged in to perform this action")

        return f(*args, **kwargs)

    return wrapper
```

File: multinet/api.py
Comment (waxlamp) [reply multinet/api.py:127]: This is actually for the typechecker. `current_user()` returns an `Optional[User]` so this assert statement is necessary to narrow the type down to `User`. The use of `@require_login` ensures that in this instance the function never returns `None`.

**Review by jjnesbitt — APPROVED** [review]
:shipit:

**Review by JackWilb — APPROVED** [review]

**Pull request merged**
